server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -1263,6 +1269,39 @@ public int hashCode()
     }
   }
 
+  private static void bindColumnValuesToQueryWithInCondition(
+      final String columnName,
+      final List<String> values,
+      final Update query
+  )
+  {
+    if (values == null) {
+      return;
+    }
+
+    for (int i = 0; i < values.size(); i++) {
+      query.bind(StringUtils.format("%s%d", columnName, i), values.get(i));
+    }
+  }
+
+  private int deletePendingSegmentsById(Handle handle, String datasource, List<String> pendingSegmentIds)
+  {
+    if (pendingSegmentIds.isEmpty()) {
+      return 0;
+    }
+
+    Update query = handle.createStatement(
```

Comment:
Should this operation be broken up into smaller batches in case there are too many segments to delete? Otherwise we might end up with a large `IN` clause.

Reply:
I am not sure but I think that a batch of single DELETEs might even be faster.
```
DELETE where id = id1;
DELETE where id = id2;
```

Reply:
I agree that batches can be too large.
Can't individual calls be too slow because of the network in between the Overlord and the Metadata store in production deployments?

Reply:
The `deleteSegments()` method also does the delete one segment ID at a time. These wouldn't really count as individual calls as you would be performing it inside a single transaction.

I assume using single DELETEs would be faster as each statement would be able to quickly find the only target row through a search of the index.

With an IN clause, depending on the underlying DB implementation, it would either be (a) find target rows corresponding to every value of the IN filter (better, but not likely) (b) scan every row and evaluate against the IN filter.

Although, I am sure every SQL database would have optimizations that make IN performance optimal. But it still does not seem reasonable to have an IN filter with more than say 100 values.

Reply:
Made changes to delete pending segments in batches of size at most 100

Reply:
thanks!